File: fsharp/hamming/Hamming.fs
module Hamming

open System

let mismatch ( (a: Char, b: Char) ) = a <> b
    
let distance (strand1: string) (strand2: string): int option =
    if strand1.Length <> strand2.Length then None
    else
        let pairs: (char * char)[] = Array.zip (strand1.ToCharArray()) (strand2.ToCharArray())
        let mismatches: (char * char)[] = Array.filter ( mismatch ) pairs
        Some mismatches.Length

File: scala/hamming/src/main/scala/Hamming.scala
object Hamming {
  def distance(dnaStrandOne: String, dnaStrandTwo: String): Option[Int] = {
      if (dnaStrandOne.length() == dnaStrandTwo.length()) {
        val pairs = dnaStrandOne.toCharArray() zip dnaStrandTwo.toCharArray()
        val matches = pairs.filter { case (a, b) => a != b }
        Some(matches.length)
      }
      else None
    }
}
